### PR TITLE
Refactor debugging through coding

### LINKED ISSUE #5133: Correct text around `InkFeature` on 'Debugging Flutter apps programmatically' page
### Page URL

https://docs.flutter.dev/testing/code-debugging

### Description

image

The `InkFeature` should be `_InkFeature`.  Because  `_InkFeature` is an implementation detail of the `Material` widget.
 
image

`_InkFeature` is a Widget ，`InkFeature` is not ，it's just a `class`.

### PR DESCRIPTION
**Pull request by atsansone**
Fixes #5133

### CODE REVIEW

File: src/testing/code-debugging.md [hunk 1]
```diff
@@ -253,7 +253,7 @@ you call `toStringDeep()` on the root of the widget tree.)
 You'll see a lot of widgets in there that don't appear in your
 application's source, because they are inserted by the framework's
 widgets' build functions. For example,
-[`InkFeature`][] is an implementation detail of the [`Material`][] widget.
+[`_InkFeature_`][] is an implementation detail of the [`Material`][] widget.
```
Comment (parlough) [inline src/testing/code-debugging.md:256]: 1. I'm assuming you meant to make this private, in which case it should be just: `[`_InkFeature`][]`

2. As far as I can tell, there is no `_InkFeature` class in Flutter. There is however a `_InkFeatures`.

   There must have been a change in Flutter at some point as the above tree doesn't include `_InkFeatures`. Can you follow the instructions above to regenerate the dumped widget tree, update that, then update this mention to the still relevant (implementation detail) widget.

File: src/testing/code-debugging.md [hunk 2]
```diff
@@ -886,41 +886,45 @@ You can set these flags either by editing the framework code,
...
-<dt markdown="1">[`debugPrintMarkNeedsPaintStacks`][]</dt>
-<dd markdown="1">Similar to `debugPrintMarkNeedsLayoutStacks`,
-    but for excess painting. You can use the `debugPrintStack()`
-    method from the `services` library to print your own stack
-    traces on demand, if this kind of approach is useful to you.
+    **Render Tree** tab.
+{% endcomment %}
+
+[`debugPaintLayerBordersEnabled`][]</dt>
+: PENDING
```
Comment (parlough) [inline src/testing/code-debugging.md:904]: If you have a chance, can you make an issue for documenting this? Otherwise it might stay as pending indefinitely.

File: src/testing/code-debugging.md [hunk 3]
```diff
@@ -886,41 +886,45 @@ You can set these flags either by editing the framework code,
 or by importing the module and setting the value in your
 `main()` method, following by a hot restart.
 
-<dl markdown="1">
-<dt markdown="1">[`debugDumpRenderTree()`][]</dt>
-<dd markdown="1">
-  Call this function when not in a layout or repaint
+[`debugDumpRenderTree()`][]</dt>
```
Comment (parlough) [inline src/testing/code-debugging.md:889]: Here and in the other cases below, you'll need to the drop the unmatched `</dt>` tags as they show up in the rendered HTML.

```suggestion
[`debugDumpRenderTree()`][]
```

**Review by parlough — COMMENTED** [review]
Thanks for this!

A few issues and comments:

File: src/testing/code-debugging.md (hunk 1 above)
Comment (atsansone) [reply src/testing/code-debugging.md:256]: It's InkFeature set to private. @HansMuller verified this for me.

Comment (HansMuller) [reply src/testing/code-debugging.md]: As @parlough points out, the current name of the class is `_InkFeatures`.

Comment (atsansone) [reply src/testing/code-debugging.md]: @HansMuller : Does this mean every instance of `_InkFeature` becomes `_InkFeatures` and I remove the link to `InkFeature`? Also, how would the tree change?

Thanks!

Comment (HansMuller) [reply src/testing/code-debugging.md]: The Material class creates an `_InkFeatures` widget and AFAICT Flutter does not contain a widget called `_InkFeature`. The tree will not change; view this update as correcting a spelling mistake.

File: src/testing/code-debugging.md [hunk 4]
```diff
@@ -167,11 +168,17 @@ class AppHome extends StatelessWidget {
...
 
-The previous app outputs something like the following
-(the precise details vary by the version of the framework,
-the size of the device, and so forth):
+This application produces the following widget tree.
 
-```
+<details markdown="1">
+<summary><strong>
+Expand the widget tree for Example 1
```
Comment (parlough) [inline src/testing/code-debugging.md:175]: Perhaps "Expand to view the" or "Expand for" would be more accurate? Since it's not the widget tree itself being expanded.

```suggestion
Expand to view the widget tree for Example 1
```

File: src/testing/code-debugging.md [hunk 5]
```diff
@@ -284,10 +293,12 @@ or an event handler is the best solution.
 To call `debugDumpRenderTree()`, you need to add `import
 'package:flutter/rendering.dart';` to your source file.
 
-The output for the previous tiny example would look something like
-the following:
+<details markdown="1">
+<summary><strong>
+Expand the render tree for Example 1
```
Comment (parlough) [inline src/testing/code-debugging.md:298]: Perhaps "Expand to view the" or "Expand for" would be more accurate? Since it's not the render tree itself being expanded.

```suggestion
Expand to view the output render tree for Example 1
```

File: src/testing/code-debugging.md [hunk 6]
```diff
@@ -644,9 +656,11 @@ objects to the method's argument, and call the superclass method.
 
 If you are trying to debug a compositing issue, you can use
 [`debugDumpLayerTree()`][].
-For the previous example, it would output:
 
-```
+<details markdown="1">
+<summary>Expand the layer tree for Example 1</summary>
```
Comment (parlough) [inline src/testing/code-debugging.md:661]: Same as above, perhaps "Expand to view the" or "Expand for" would be more accurate? Since it's not the layer tree itself being expanded.

```suggestion
<summary>Expand to view the output layer tree for Example 1</summary>
```

File: src/testing/code-debugging.md [hunk 7]
```diff
@@ -677,12 +693,18 @@ used to reduce how much needs to be repainted.
 
 ### Focus tree
 
-To debug a focus or shortcut issue, you can dump the focus tree
-using [`debugDumpFocusTree()`][].
+To debug a focus or shortcut issue, dump the focus tree
+using the [`debugDumpFocusTree()`][] method.
 
-For example:
+#### Example 4: Focus tree dump
```
Comment (parlough) [inline src/testing/code-debugging.md:699]: Why does this have a header and where does the 4 come from? The previous dumps didn't have a header.

File: src/testing/code-debugging.md [hunk 8]
```diff
@@ -677,12 +693,18 @@ used to reduce how much needs to be repainted.
...
+{:.no_toc}
 
-```
+The `debugDumpFocusTree()` method returns the focus tree for the app.
```
Comment (parlough) [inline src/testing/code-debugging.md:702]: According to the [docs](https://api.flutter.dev/flutter/widgets/debugDumpFocusTree.html), it doesn't return the tree, it just prints it.

Also, doesn't this text just duplicate the content on L696-L697. Could it just be dropped?

File: src/testing/code-debugging.md [hunk 9]
```diff
@@ -677,12 +693,18 @@ used to reduce how much needs to be repainted.
...
+
+<details markdown="1">
+<summary>Expand the focus tree for Example 1</summary>
```
Comment (parlough) [inline src/testing/code-debugging.md:705]:
```suggestion
<summary>Expand to view the focus tree for Example 1</summary>
```

File: src/testing/code-debugging.md [hunk 10]
```diff
@@ -758,22 +782,28 @@ extensive logging when the focus changes.
...
-enabling a system accessibility tool or the `SemanticsDebugger`.
+To obtain a dump of the Semantics tree
+(the tree presented to the system accessibility APIs), use the
+[`debugDumpSemanticsTree()`][] method.
+To use this, first enable accessibility.
+To enable accessibility, use a system accessibility tool
+or the `SemanticsDebugger`.
 
-For the previous example, it would output the following:
+The `debugDumpSemanticsTree()` method returns the semantic tree for the app.
```
Comment (parlough) [inline src/testing/code-debugging.md:792]: It doesn't return the tree, just prints it.

```suggestion
The `debugDumpSemanticsTree()` function prints the semantic tree for the app.
```

File: src/testing/code-debugging.md [hunk 11]
```diff
@@ -677,12 +693,18 @@ used to reduce how much needs to be repainted.
 
 ### Focus tree
 
-To debug a focus or shortcut issue, you can dump the focus tree
-using [`debugDumpFocusTree()`][].
+To debug a focus or shortcut issue, dump the focus tree
+using the [`debugDumpFocusTree()`][] method.
```
Comment (parlough) [inline src/testing/code-debugging.md:697]: Since it's not a class or instance function, it is not a method, just a function.

```suggestion
using the [`debugDumpFocusTree()`][] function.
```

File: src/testing/code-debugging.md [hunk 12]
```diff
@@ -274,8 +283,8 @@ widget's description.
 ### Render tree
 
 If you are trying to debug a layout issue, then the Widgets layer's
-tree might be insufficiently detailed. In that case, you can dump the
-rendering tree by calling [`debugDumpRenderTree()`][].
+tree might be insufficiently detailed. To dump the rendering tree,
+call the [`debugDumpRenderTree()`][] method.
```
Comment (parlough) [inline src/testing/code-debugging.md:287]: Since it's not a class or instance function, it is not a method, just a function.

```suggestion
call the [`debugDumpRenderTree()`][] function.
```

File: src/testing/code-debugging.md [hunk 13]
```diff
@@ -783,9 +813,9 @@ begin/end, you can toggle the [`debugPrintBeginFrameBanner`][]
 and the [`debugPrintEndFrameBanner`][] booleans to print the
 beginning and end of the frames to the console.
 
-For example:
+**The print frame banner log for Example 1**
```
Comment (parlough) [inline src/testing/code-debugging.md:816]: We don't really use this style of introducing a block of code anywhere else, perhaps stick to introductory text and a colon for familiarity?

File: src/testing/code-debugging.md [hunk 14]
```diff
@@ -802,7 +832,12 @@ You can also debug a layout problem visually, by setting
 This is a boolean from the `rendering` library. It can be
 enabled at any time and affects all painting while it is true.
 The easiest way to set it is at the top of your `void main()`
-entry point. See an example in the following code:
+entry point.
+
+#### Example 2
```
Comment (parlough) [inline src/testing/code-debugging.md:837]: Calling this "Example 2" was a bit confusing to me. Example 2 of what? It doesn't seem to be related to "Example 1".

File: src/testing/code-debugging.md [hunk 15]
```diff
@@ -969,8 +1008,11 @@ setting the `showPerformanceOverlay` property to `true` on the
 [`MaterialApp`][], [`CupertinoApp`][], or [`WidgetsApp`][]
 constructor:
 
+#### Example 3
+{:.no_toc}
+
 <?code-excerpt "lib/performance_overlay.dart (PerfOverlay)" replace="/showPerformanceOverlay: true,/[[highlight]]$&[[\/highlight]]/g"?>
-{% prettify dart %}
+```dart
```
Comment (parlough) [inline src/testing/code-debugging.md:1015]: Code highlighting doesn't yet working without prettify (I'm working on this now), so you'll have to switch this one back.

The rendered output having the highlight tag

File: src/testing/code-debugging.md [hunk 16]
```diff
@@ -988,14 +1030,14 @@ class MyApp extends StatelessWidget {
...
-{% endprettify %}
+```
 
 (If you're not using `MaterialApp`, `CupertinoApp`,
 or `WidgetsApp`, you can get the same effect by wrapping your
 application in a stack and putting a widget on your stack that was
 created by calling [`PerformanceOverlay.allEnabled()`][].)
 
-For information on how to interpret the graphs in the overlay,
+TO learn how to interpret the graphs in the overlay,
```
Comment (parlough) [inline src/testing/code-debugging.md:1040]:
```suggestion
To learn how to interpret the graphs in the overlay,
```

File: src/testing/code-debugging.md [hunk 17]
```diff
@@ -878,49 +913,53 @@ result in asserts and generally interfere with your efforts.
...
+  If any static widgets are rotating through the colors of the rainbow,
+  consider adding repaint boundaries to those areas.
+
+[`debugPrintMarkNeedsLayoutStacks`][]
+: Enable this flag if you're seeing more layouts
+  than you expect (for example, on the timeline, on a profile,
+  or from a `print` statement inside a layout method).
+  When enabled, the console shows stack traces that explain why each
+  the inspector marked each render object as dirty for layout.
+  You can use the `debugPrintStack()` method from the
```
Comment (parlough) [inline src/testing/code-debugging.md:954]:
```suggestion
  You can use the `debugPrintStack()` function from the
```

File: src/testing/code-debugging.md [hunk 18]
```diff
@@ -878,49 +913,53 @@ result in asserts and generally interfere with your efforts.
...
+  `services` library to print your own stack traces on demand,
+  if this helps you.
```
Comment (parlough) [inline src/testing/code-debugging.md:956]: Since you're touching this, consider dropping the conditional or putting it first.

File: src/testing/code-debugging.md [hunk 19]
```diff
@@ -878,49 +913,53 @@ result in asserts and generally interfere with your efforts.
...
+
+[`debugPrintMarkNeedsPaintStacks`][]
+: Similar to `debugPrintMarkNeedsLayoutStacks`,
+  but for excess painting. You can use the `debugPrintStack()`
+  method from the `services` library to print your own stack
```
Comment (parlough) [inline src/testing/code-debugging.md:961]:
```suggestion
  function from the `services` library to print your own stack
```

File: src/testing/code-debugging.md [hunk 20]
```diff
@@ -878,49 +913,53 @@ result in asserts and generally interfere with your efforts.
...
+  traces on demand, if this kind of approach is useful to you.
```
Comment (parlough) [inline src/testing/code-debugging.md:962]: Since you're touching this, consider dropping the conditional or putting it first.

File: src/testing/code-debugging.md [hunk 21]
```diff
@@ -878,49 +913,53 @@ result in asserts and generally interfere with your efforts.
...
+: To enable this flag in the Flutter inspector, click **Highlight Repaints**.
+  If any static widgets are rotating through the colors of the rainbow,
+  consider adding repaint boundaries to those areas.
+
+[`debugPrintMarkNeedsLayoutStacks`][]
+: Enable this flag if you're seeing more layouts
+  than you expect (for example, on the timeline, on a profile,
+  or from a `print` statement inside a layout method).
+  When enabled, the console shows stack traces that explain why each
+  the inspector marked each render object as dirty for layout.
```
Comment (parlough) [inline src/testing/code-debugging.md:953]: Looks like the wording got messed up here. 

```suggestion
  When enabled, the framework outputs stack traces to the console
  to explain why each render object is being marked as needing layout.
```

**Review by parlough — COMMENTED** [review]
Thanks for those adjustments! 

A few more comments:

**Comment by atsansone** [comment]
@HansMuller : I'm going to set up time to discuss this. The issue raised covers one concern but I've found more.

**Comment by domesticmouse** [comment]
This needs maintenance to bring it up to date with `main`

**Comment by domesticmouse** [comment]
Looks like the CLA bot is unhappy with this PR. The author `30806939+DemX86@users.noreply.github.com` is causing issues.

**Comment by parlough** [comment]
@atsansone Let's talk about this tomorrow when we meet and get this landed?

**Comment by domesticmouse** [comment]
Please update excerpts. CI is unhappy.

**Comment by domesticmouse** [comment]
Excerpts are still cranky

```console
diff --no-dereference --exclude '_*' --color -U2 -r /tmp/tmp.HhotR5veCO/src/testing/code-debugging.md /app/src/testing/code-debugging.md
--- /tmp/tmp.HhotR5veCO/src/testing/code-debugging.md	2023-11-03 11:33:53.862819465 -0700
+++ /app/src/testing/code-debugging.md	2023-11-03 11:35:44.265558144 -0700
@@ -361,7 +361,7 @@
         child: TextButton(
           onPressed: () {
-            debugDumpApp();
+            debugDumpLayerTree();
           },
-          child: const Text('Dump Widget Tree'),
+          child: const Text('Dump Layer Tree'),
         ),
       ),
=> ERROR: The above code excerpts needed to be updated!
```

**Review by sfshaza2 — COMMENTED** [review]
I reviewed this yesterday but, for some reason, didn't submit it.

File: src/testing/code-debugging.md [hunk 22]
```diff
@@ -113,28 +117,36 @@ void someFunction(double offset) {
...
+Each layer of the Flutter framework provides a function to dump its
+current state or events to the console using the `debugPrint` property.
 
-{% comment %}  DevTool's doesn't currently print the render tree.
 {{site.alert.note}}
-  If you use [DevTools][] and its [Flutter inspector][] to
-  view a visual layout of the render tree, you probably won't
-  need to use these text-based dump tools.
+  All of the following examples were run as macOS native apps on
+  a MacBook Pro M1 Pro. These will differ from any dumps your
```
[the remark this answers is not in the document]
Comment (sfshaza2) [reply src/testing/code-debugging.md:127]: Still not fixed.

**Review by sfshaza2 — APPROVED** [review]
Once @parlough is happy, lgtm. (There were some minor issues I pointed out in my last review that are still present.) I'm approving just to shortcut the process.

File: src/testing/code-debugging.md [hunk 23]
```diff
@@ -17,7 +17,7 @@ For a full list of debugging and profiling tools, see the
 
 [`flutter_gdb`]: https://github.com/flutter/engine/blob/main/sky/tools/flutter_gdb
 
-## Logging
+## Log your application
```
Comment (domesticmouse) [inline src/testing/code-debugging.md:20]: Should this be "Add logging to your application"?

I don't think of log as a verb outside of tree felling...

File: src/testing/code-debugging.md (hunk 13 above)
Comment (atsansone) [reply src/testing/code-debugging.md:541]: Stet. This is consistent with the other log listings, which are behind bolded accordions.

File: src/testing/code-debugging.md [hunk 24]
```diff
@@ -815,25 +574,23 @@ void main() {
...
-(from widgets like `Center` and `Align`) is shown with
-yellow arrows, and spacers (from widgets like
-`Container` when they have no child) are shown in gray.
+When enabled, Flutter displays the following changes to your app:
+
+* all boxes in a bright teal border
+* all padding, from widgets like `Padding`,
+  as a box with a faded blue fill and blue border around the child widget
+* all alignment, from widgets like `Center` and `Align`, with yellow arrows
+* all spacers, from widgets like `Container` when they have no child, in gray
```
[the remark this answers is not in the document]
Comment (atsansone) [reply src/testing/code-debugging.md:583]: Made other changes.

File: src/testing/code-debugging.md [hunk 25]
```diff
@@ -843,18 +600,17 @@ or the [`debugRepaintRainbowEnabled`][] flag,
...
+## Debug animation issues
 
 {{site.alert.note}}
-  The easiest way to debug animations is to slow them down.
-  You can do this using the **Slow Animations** button in
-  DevTools' [Inspector view][], which slows down the
-  animation by 5x. If you want more control over the
-  amount of slowness, use the following instructions.
+  To debug animations with the least effort, slow them down.
+  To slow down the animation, click **Slow Animations** in DevTools' [Inspector view][]. This reduces the animation to 20% speed.
```
[the remark this answers is not in the document]
Comment (atsansone) [reply src/testing/code-debugging.md:611]: It removes the value judgment of how easy this is to do.

File: src/testing/code-debugging.md (hunk 15 above)
Comment (atsansone) [reply src/testing/code-debugging.md:738]: Removed.

**Review by sfshaza2 — APPROVED** [review]
lgtm

**Review by domesticmouse — APPROVED** [review]
